File: trunk/NowPlaying/iPhone/Classes/Movie.m
// Copyright 2008 Cyrus Najmabadi
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "Movie.h"

#import "Utilities.h"

@interface Movie()
@property (copy) NSString* identifier;
@property (copy) NSString* canonicalTitle;
@property (copy) NSString* displayTitle;
@property (copy) NSString* rating;
@property NSInteger length;
@property (copy) NSString* imdbAddress;
@property (copy) NSString* poster;
@property (copy) NSString* synopsis;
@property (copy) NSString* studio;
@property (retain) NSDate* releaseDate;
@property (retain) NSArray* directors;
@property (retain) NSArray* cast;
@property (retain) NSArray* genres;
@property (retain) NSString* cachedRatingAndRuntimeString;
@end


@implementation Movie

property_definition(identifier);
property_definition(canonicalTitle);
property_definition(displayTitle);
property_definition(rating);
property_definition(length);
property_definition(releaseDate);
property_definition(imdbAddress);
property_definition(poster);
property_definition(synopsis);
property_definition(studio);
property_definition(directors);
property_definition(cast);
property_definition(genres);
@synthesize cachedRatingAndRuntimeString;

- (void) dealloc {
    self.identifier = nil;
    self.canonicalTitle = nil;
    self.rating = nil;
    self.length = 0;
    self.releaseDate = nil;
    self.imdbAddress = nil;
    self.poster = nil;
    self.synopsis = nil;
    self.displayTitle = nil;
    self.studio = nil;
    self.directors = nil;
    self.cast = nil;
    self.genres = nil;
    self.cachedRatingAndRuntimeString = nil;

    [super dealloc];
}


static NSString* articles[] = {
@"Der", @"Das", @"Ein", @"Eine", @"The",
@"A", @"An", @"La", @"Las", @"Le",
@"Les", @"Los", @"El", @"Un", @"Une",
@"Una", @"Il", @"O", @"Het", @"De",
@"Os", @"Az", @"Den", @"Al", @"En",
@"L'"
};

+ (NSString*) makeCanonical:(NSString*) title {
    title = [title stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];

    for (int i = 0; i < ArrayLength(articles); i++) {
        NSString* article = articles[i];
        if ([title hasSuffix:[NSString stringWithFormat:@", %@", article]]) {
            return [NSString stringWithFormat:@"%@ %@", article, [title substringToIndex:(title.length - article.length - 2)]];
        }
    }

    return title;
}


+ (NSString*) makeDisplay:(NSString*) title {
    title = [title stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];

    for (int i = 0; i < ArrayLength(articles); i++) {
        NSString* article = articles[i];
        if ([title hasPrefix:[NSString stringWithFormat:@"%@ ", article]]) {
            return [NSString stringWithFormat:@"%@, %@", [title substringFromIndex:(article.length + 1)], article];
        }
    }

    return title;
}


- (id) initWithIdentifier:(NSString*) identifier_
           canonicalTitle:(NSString*) canonicalTitle_
             displayTitle:(NSString*) displayTitle_
                   rating:(NSString*) rating_
                   length:(NSInteger) length_
              releaseDate:(NSDate*) releaseDate_
              imdbAddress:(NSString*) imdbAddress_
                   poster:(NSString*) poster_
                 synopsis:(NSString*) synopsis_
                   studio:(NSString*) studio_
                directors:(NSArray*) directors_
                     cast:(NSArray*) cast_
                   genres:(NSArray*) genres_ {
    if (self = [self init]) {
        self.identifier = [Utilities nonNilString:identifier_];
        self.canonicalTitle = [Utilities nonNilString:canonicalTitle_];
        self.displayTitle = [Utilities nonNilString:displayTitle_];
        self.rating = [Utilities nonNilString:rating_];
        self.length = length_;
        self.releaseDate = releaseDate_;
        self.imdbAddress = [Utilities nonNilString:imdbAddress_];
        self.poster = [Utilities nonNilString:poster_];
        self.synopsis = [Utilities nonNilString:synopsis_];
        self.studio = [Utilities nonNilString:studio_];
        self.directors = [Utilities nonNilArray:directors_];
        self.cast = [Utilities nonNilArray:cast_];
        self.genres = [Utilities nonNilArray:genres_];
    }

    return self;
}


+ (Movie*) movieWithIdentifier:(NSString*) identifier
                         title:(NSString*) title
                        rating:(NSString*) rating
                        length:(NSInteger) length
                   releaseDate:(NSDate*) releaseDate
                   imdbAddress:(NSString*) imdbAddress
                        poster:(NSString*) poster
                      synopsis:(NSString*) synopsis
                        studio:(NSString*) studio
                     directors:(NSArray*) directors
                          cast:(NSArray*) cast
                        genres:(NSArray*) genres {
    rating = [rating stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
    if (rating.length == 0) {
        rating = @"NR";
    }

    return [[[Movie alloc] initWithIdentifier:identifier
                               canonicalTitle:[self makeCanonical:title]
                                 displayTitle:[self makeDisplay:title]
                                       rating:rating
                                       length:length
                                  releaseDate:releaseDate
                                  imdbAddress:imdbAddress
                                       poster:poster
                                     synopsis:[Utilities stripHtmlCodes:synopsis]
                                       studio:studio
                                    directors:directors
                                         cast:cast
                                       genres:genres] autorelease];
}


+ (Movie*) movieWithDictionary:(NSDictionary*) dictionary {
    return [[[Movie alloc] initWithIdentifier:[dictionary objectForKey:identifier_key]
                               canonicalTitle:[dictionary objectForKey:canonicalTitle_key]
                                 displayTitle:[dictionary objectForKey:displayTitle_key]
                                       rating:[dictionary objectForKey:rating_key]
                                       length:[[dictionary objectForKey:length_key] intValue]
                                  releaseDate:[dictionary objectForKey:releaseDate_key]
                                  imdbAddress:[dictionary objectForKey:imdbAddress_key]
                                       poster:[dictionary objectForKey:poster_key]
                                     synopsis:[dictionary objectForKey:synopsis_key]
                                       studio:[dictionary objectForKey:studio_key]
                                    directors:[dictionary objectForKey:directors_key]
                                         cast:[dictionary objectForKey:cast_key]
                                       genres:[dictionary objectForKey:genres_key]] autorelease];
}


- (NSDictionary*) dictionary {
    NSMutableDictionary* dictionary = [NSMutableDictionary dictionary];
    [dictionary setValue:identifier                         forKey:identifier_key];
    [dictionary setValue:canonicalTitle                     forKey:canonicalTitle_key];
    [dictionary setValue:displayTitle                       forKey:displayTitle_key];
    [dictionary setValue:rating                             forKey:rating_key];
    [dictionary setValue:[NSNumber numberWithInt:length]    forKey:length_key];
    [dictionary setValue:releaseDate                        forKey:releaseDate_key];
    [dictionary setValue:imdbAddress                        forKey:imdbAddress_key];
    [dictionary setValue:poster                             forKey:poster_key];
    [dictionary setValue:synopsis                           forKey:synopsis_key];
    [dictionary setValue:studio                             forKey:studio_key];
    [dictionary setValue:directors                          forKey:directors_key];
    [dictionary setValue:cast                               forKey:cast_key];
    [dictionary setValue:genres                             forKey:genres_key];
    return dictionary;
}


- (NSString*) description {
    return self.dictionary.description;
}


- (BOOL) isEqual:(id) anObject {
    Movie* other = anObject;

    return [canonicalTitle isEqual:other.canonicalTitle];
}


- (NSUInteger) hash {
    return canonicalTitle.hash;
}


- (id) copyWithZone:(NSZone*) zone {
    return [self retain];
}


- (BOOL) isUnrated {
    return rating.length == 0 || [rating isEqual:@"NR"] || [rating isEqual:@"Not Rated"];
}


- (NSString*) ratingString {
    if (self.isUnrated) {
        return NSLocalizedString(@"Unrated", nil);
    }  else {
        return [NSString stringWithFormat:NSLocalizedString(@"Rated %@", nil), rating];
    }
}


- (NSString*) runtimeString {
    NSString* hoursString = @"";
    NSString* minutesString = @"";

    if (length > 0) {
        NSInteger hours = length / 60;
        NSInteger minutes = length % 60;

        if (hours == 1) {
            hoursString = NSLocalizedString(@"1 hour", nil);
        } else if (hours > 1) {
            hoursString = [NSString stringWithFormat:NSLocalizedString(@"%d hours", nil), hours];
        }

        if (minutes == 1) {
            minutesString = NSLocalizedString(@"1 minute", nil);
        } else if (minutes > 1) {
            minutesString = [NSString stringWithFormat:NSLocalizedString(@"%d minutes", nil), minutes];
        }
    }

    return [NSString stringWithFormat:NSLocalizedString(@"%@ %@", "2 hours 34 minutes"), hoursString, minutesString];
}


- (NSString*) ratingAndRuntimeString {
    if (cachedRatingAndRuntimeString == nil) {
        self.cachedRatingAndRuntimeString = 
        [NSString stringWithFormat:NSLocalizedString(@"%@. %@", "Rated R. 2 hours 34 minutes"), self.ratingString, self.runtimeString];
    }
    
    return cachedRatingAndRuntimeString;
}

@end

File: tags/NowPlaying/iPhone/3.8.2/NowPlaying/Android/src/org/metasyntactic/data/Movie.java
//Copyright 2008 Cyrus Najmabadi
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.
package org.metasyntactic.data;

import android.os.Parcel;
import android.os.Parcelable;
import org.metasyntactic.io.AbstractPersistable;
import org.metasyntactic.io.PersistableInputStream;
import org.metasyntactic.io.PersistableOutputStream;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;

public class Movie extends AbstractPersistable implements Parcelable, Comparable<Movie> {
  private static final long serialVersionUID = 4570788252867866289L;
  private final String identifier;
  private final String canonicalTitle;
  private final String displayTitle;
  private final String rating;
  private final int length; // minutes;
  private final String imdbAddress;
  private final Date releaseDate;
  private final String poster;
  private final String synopsis;
  private final String studio;
  private final List<String> directors;
  private final List<String> cast;
  private final List<String> genres;

  public void persistTo(final PersistableOutputStream out) throws IOException {
    out.writeString(identifier);
    out.writeString(canonicalTitle);
    out.writeString(displayTitle);
    out.writeString(rating);
    out.writeInt(length);
    out.writeString(imdbAddress);
    out.writeDate(releaseDate);
    out.writeString(poster);
    out.writeString(synopsis);
    out.writeString(studio);
    out.writeStringCollection(directors);
    out.writeStringCollection(cast);
    out.writeStringCollection(genres);
  }

  public static final Reader<Movie> reader = new AbstractReader<Movie>() {
    public Movie read(final PersistableInputStream in) throws IOException {
      final String identifier = in.readString();
      final String canonicalTitle = in.readString();
      final String displayTitle = in.readString();
      final String rating = in.readString();
      final int length = in.readInt();
      final String imdbAddress = in.readString();
      final Date releaseDate = in.readDate();
      final String poster = in.readString();
      final String synopsis = in.readString();
      final String studio = in.readString();
      final List<String> directors = in.readStringList();
      final List<String> cast = in.readStringList();
      final List<String> genres = in.readStringList();

      return new Movie(identifier, canonicalTitle, displayTitle, rating, length, imdbAddress, releaseDate, poster, synopsis, studio, directors, cast,
        genres);
    }
  };

  private Movie(final String identifier, final String canonicalTitle, final String displayTitle, final String rating, final int length,
    final String imdbAddress, final Date releaseDate, final String poster, final String synopsis, final String studio, final List<String> directors,
    final List<String> cast, final List<String> genres) {
    this.identifier = identifier;
    this.canonicalTitle = canonicalTitle;
    this.rating = rating;
    this.length = length;
    this.imdbAddress = imdbAddress;
    this.releaseDate = releaseDate;
    this.poster = poster;
    this.synopsis = synopsis;
    this.studio = studio;
    this.directors = directors;
    this.cast = cast;
    this.genres = genres;
    this.displayTitle = displayTitle;
  }

  public Movie(final String identifier, final String title, final String rating, final int length, final String imdbAddress, final Date releaseDate,
    final String poster, final String synopsis, final String studio, final List<String> directors, final List<String> cast,
    final List<String> genres) {
    this(identifier, makeCanonical(title), makeDisplay(title), rating, length, imdbAddress, releaseDate, poster, synopsis, studio, directors, cast,
      genres);
  }

  public String getIdentifier() {
    return identifier;
  }

  public String getCanonicalTitle() {
    return canonicalTitle;
  }

  public String getDisplayTitle() {
    return displayTitle;
  }

  public String getRating() {
    return rating;
  }

  public int getLength() {
    return length;
  }

  public String getIMDbAddress() {
    return imdbAddress;
  }

  public Date getReleaseDate() {
    return releaseDate;
  }

  public String getPoster() {
    return poster;
  }

  public String getSynopsis() {
    return synopsis;
  }

  public String getStudio() {
    return studio;
  }

  public List<String> getDirectors() {
    return Collections.unmodifiableList(directors);
  }

  public List<String> getCast() {
    return Collections.unmodifiableList(cast);
  }

  public List<String> getGenres() {
    return Collections.unmodifiableList(genres);
  }

  @Override
  public boolean equals(final Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof Movie)) {
      return false;
    }

    final Movie movie = (Movie)o;

    if (canonicalTitle != null ? !canonicalTitle.equals(movie.canonicalTitle) : movie.canonicalTitle != null) {
      return false;
    }

    return true;
  }

  @Override
  public int hashCode() {
    return canonicalTitle != null ? canonicalTitle.hashCode() : 0;
  }

  @Override
  public String toString() {
    return canonicalTitle;
  }

  private static final String[] prefixArticles;
  private static final String[] suffixArticles;

  static {
    final String[] articles = {"Der", "Das", "Ein", "Eine", "The", "A", "An", "La", "Las", "Le", "Les", "Los", "El", "Un", "Une", "Una", "Il", "O", "Het", "De", "Os", "Az", "Den", "Al", "En", "L'"};

    prefixArticles = new String[articles.length];
    suffixArticles = new String[articles.length];

    for (int i = 0; i < articles.length; i++) {
      prefixArticles[i] = articles[i] + ' ';
      suffixArticles[i] = ", " + articles[i];
    }
  }

  public static String makeCanonical(final String title) {
    for (int i = 0; i < prefixArticles.length; i++) {
      final String prefixArticle = prefixArticles[i];
      final String suffixArticle = suffixArticles[i];

      if (title.endsWith(suffixArticle)) {
        return prefixArticle + title.substring(0, title.length() - suffixArticle.length());
      }
    }

    return title;
  }

  public static String makeDisplay(final String title) {
    for (int i = 0; i < prefixArticles.length; i++) {
      final String prefixArticle = prefixArticles[i];
      final String suffixArticle = suffixArticles[i];

      if (title.startsWith(prefixArticle)) {
        return title.substring(prefixArticle.length()) + suffixArticle;
      }
    }

    return title;
  }

  public int describeContents() {
    return 0;
  }

  public void writeToParcel(final Parcel dest, final int flags) {
    dest.writeString(identifier);
    dest.writeString(canonicalTitle);
    dest.writeString(displayTitle);
    dest.writeString(rating);
    dest.writeInt(length);
    dest.writeString(imdbAddress);
    dest.writeValue(releaseDate);
    dest.writeString(poster);
    dest.writeString(synopsis);
    dest.writeString(studio);
    dest.writeStringList(directors);
    dest.writeStringList(cast);
    dest.writeStringList(genres);
  }

  public static final Creator<Movie> CREATOR = new Creator<Movie>() {
    public Movie createFromParcel(final Parcel source) {
      final String identifier = source.readString();
      final String canonicalTitle = source.readString();
      final String displayTitle = source.readString();
      final String rating = source.readString();
      final int length = source.readInt();
      final String imdbAddress = source.readString();
      final Date releaseDate = (Date)source.readValue(null);
      final String poster = source.readString();
      final String synopsis = source.readString();
      final String studio = source.readString();
      final List<String> directors = new ArrayList<String>();
      source.readStringList(directors);
      final List<String> cast = new ArrayList<String>();
      source.readStringList(cast);
      final List<String> genres = new ArrayList<String>();
      source.readStringList(genres);

      return new Movie(identifier, canonicalTitle, displayTitle, rating, length, imdbAddress, releaseDate, poster, synopsis, studio, directors, cast,
        genres);
    }

    public Movie[] newArray(final int size) {
      return new Movie[size];
    }
  };

  public int compareTo(final Movie movie) {
    return getCanonicalTitle().compareTo(movie.getCanonicalTitle());
  }

  public static final Comparator<Movie> TITLE_ORDER = new Comparator<Movie>() {
    public int compare(final Movie m1, final Movie m2) {
      return m1.getDisplayTitle().compareTo(m2.getDisplayTitle());
    }
  };
  public static final Comparator<Movie> RELEASE_ORDER = new Comparator<Movie>() {
    public int compare(final Movie m1, final Movie m2) {
      final Calendar c1 = Calendar.getInstance();
      c1.set(1900, 11, 11);
      Date d1 = c1.getTime();
      Date d2 = c1.getTime();
      if (m1.getReleaseDate() != null) {
        d1 = m1.getReleaseDate();
      }
      if (m2.getReleaseDate() != null) {
        d2 = m2.getReleaseDate();
      }
      return d2.compareTo(d1);
    }
  };
}
